grammar.js

```diff
@@ -536,6 +536,11 @@ grammar({
 
     do_clause: $ => seq(
       'do',
+      choice(
+        $._terminator, 
+        seq($._expression, $._terminator),
+        seq($.bare_tuple_expression, $._terminator),
```

Comment:
I don't think we should use `expression` and `bare_tuple_expression` in this position. I think these nodes between  the `do` and the `terminator` are *parameters* to the do block, right? So they introduce new names into a local scope.

I would create a `do_parameters` node that just contains one or more `identifier`s, separated by commas. Do you think that will work?

Reply:
The terminators are also optional, except in the nullary case.
Both `map(1:10) do x x end` and `map(1:10, 11:20) do x, y x+y end` is valid Julia syntax. The `_terminator` should be handled by `_expression_list`, that's how it's done in the Julia parser.

Reply:
@maxbrunsfeld Unfortunately, I don't believe that will work. See the following Julia program:
```
function f(g)
  println(g(1, (2, 3)))
end

f() do x, (y, z)
  x + y + z
end
```
Or another, related example:
```

function f(g)
  println(g(1, (3)))
end 

f() do x, (y::Integer)
  x + y
end
```
These programs print `6` and `4` when run, respectively.

Truth be told, I have absolutely no idea what is allowed as a parameter to a `do`-block, but it isn't as simple as comma-separated identifiers. That's why I have elected to err on the side of being permissive, to allow all valid Julia examples.

@ChrHorn, you make a good point. I would still not like to leave it as-is and leave it to the `_expression_list`, however, as for one thing it cannot express the case where you start with a terminator, and also because the arguments to the `do`-block are semantically significant, in my opinion. I will probably just revise all these terminators to be optional.

Reply:
This is what I came up with:
```javascript
do_clause: $ => seq(
  'do',
  choice(
    $._terminator,
    alias($._do_parameters, $.parameter_list)
  ),
  optional($._expression_list),
  'end'
),

_do_parameters: $ => prec.left(seq(
  sep1(',', choice($.identifier, $.tuple_parameters))
)),

tuple_parameters: $ => seq(
  '(',
  sep1(',', choice($.identifier, $.tuple_parameters)),
  ')'
),
```
This seems to handle all the cases, including tuple destructuring, and uses the `parameter_list` as scope. All existing queries that use parameters should work automatically. A `tuple_parameter` is currently also missing for the normal functions and can also be added there.